### ISSUE TITLE
Hermetic sandbox fails to move output artifacts if one of them is a read-only directory

### ISSUE THREAD

**Issue by jkeljo** [bazelbuild/bazel#20032]
### Description of the bug:

If a rule produces as one of its output artifacts a directory that does not have write permissions, building that rule in the hermetic sandbox fails with "I/O exception during sandboxed execution: Could not move output artifacts from sandboxed execution".

### Which category does this issue belong to?

Core

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

1. Install Bazelisk
2. Clone https://github.com/jkeljo/bazel_sandbox_readonly_output_directory_bug
3. Run bazelisk build -s //:test

### Which operating system are you running Bazel on?

Ubuntu 20.04

### What is the output of `bazel info release`?

release 6.4.0 (also reproes with 7.0.0-pre.20231011.2)

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
https://github.com/jkeljo/bazel_sandbox_readonly_output_directory_bug
master
fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
d9d5423e1bf8d5a660bde19e8a24a040b02cf368
```

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

no

### Have you found anything relevant by searching the web?

no

**Comment by iancha1992** [comment]
I was able to repro this error:

```
INFO: Analyzed target //:test (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
SUBCOMMAND: # //:test [action 'MakeReadonlyDirectory test/dir', configuration: eae08a0147418bbed67463ae5f68913a92ab5c68abe9a3a6ea98eee64dbbdd8a, execution platform: @local_config_platform//:host]
(cd /usr/local/google/home/heec/.cache/bazel/_bazel_heec/371fff5886a117a02b179fcac29ff83b/execroot/__main__ && \
  exec env - \
  /bin/bash -c 'mkdir -p $1; echo foo > $1/foo.txt; echo bar > $1/bar.txt; chmod -w $1' '' bazel-out/k8-fastbuild/bin/test/dir)
# Configuration: eae08a0147418bbed67463ae5f68913a92ab5c68abe9a3a6ea98eee64dbbdd8a
# Execution platform: @local_config_platform//:host
ERROR: /usr/local/google/home/heec/projects/bazel_sandbox_readonly_output_directory_bug/BUILD:3:31: MakeReadonlyDirectory test/dir failed: I/O exception during sandboxed execution: Could not move output artifacts from sandboxed execution
Target //:test failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.184s, Critical Path: 0.04s
INFO: 2 processes: 2 internal.
FAILED: Build did NOT complete successfully
```
just FYI.

**Comment by tjgq** [comment]
I'm working on a fix.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.0.0 RC5](https://releases.bazel.build/7.0.0/rc5/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
